oxbow/src/bcf.rs

```diff
@@ -0,0 +1,206 @@
+use arrow::array::{
+    ArrayRef, Float32Builder, GenericStringBuilder, Int32Builder,
+    StringArray, StringDictionaryBuilder,
+};
+use arrow::{
+    datatypes::Int32Type, error::ArrowError, record_batch::RecordBatch,
+};
+use byteorder::{LittleEndian, ReadBytesExt};
+use noodles::core::Region;
+use noodles::bcf::header::StringMaps;
+use noodles::{bcf, bgzf, csi, vcf};
+use std::sync::Arc;
+use std::{ffi::CStr, io};
+use std::io::{Read};
+
+use crate::batch_builder::{write_ipc, BatchBuilder};
+
+type BufferedReader = std::io::BufReader<std::fs::File>;
+
+
+fn read_u8<R>(reader: &mut R) -> io::Result<u8>
+where
+    R: Read,
+{
+    let mut buf = [0; 1];
+    reader.read_exact(&mut buf)?;
+    Ok(buf[0])
+}
+
+fn read_raw_header<R>(reader: &mut R) -> io::Result<String>
+where
+    R: Read,
+{
+    let l_text = reader.read_u32::<LittleEndian>().and_then(|n| {
+        usize::try_from(n).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    })?;
+
+    let mut buf = vec![0; l_text];
+    reader.read_exact(&mut buf)?;
+
+    CStr::from_bytes_with_nul(&buf)
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+        .and_then(|c_header| {
+            c_header
+                .to_str()
+                .map(|s| s.into())
+                .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+        })
+}
+
+fn read_bcf_header<R>(reader: &mut R) -> io::Result<(vcf::Header, bcf::header::StringMaps)>
+where
+    R: Read,
+{
+    // read the MAGIC number
+    let mut buf = [0; 3];
```

Comment:
Perhaps it's the allocation of size 3 that avoids the "null byte at position 3" error?

Reply:
Yeah, I'm thinking that maybe it was reading through the raw gzip stream, rather than the uncompressed stream, which may have had a null byte at that position, so reading the magic `b"BCF"` failed.

If you check the file I used, it has a correct magic number in the uncompressed stream.